# 1. Introduction 

The goal of this project is to find the best possible option to buy from the list of used cars listed on ebay Kleinanzeigen, a classifieds section of the German eBay website.

Data dictionary for the dataset:

- dateCrawled - When the ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing.
- price - The listed selling price of the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The type of vehicle.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The type of transmission.
- powerPS - The power of the car in PS.
- model - The car model name.
- odometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

In [10]:
import pandas as pd
import numpy as np

autos = pd.read_csv('Downloads/autos.csv', encoding = 'latin1') # reading the csv data into a pandas dataframe.

In [11]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [5]:
autos.info()      #Metadata of autos datafram
autos.head(10)    #First 10 rows of the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

dateCrawled                                               name  \
0  2016-03-24 11:52:17                                         Golf_3_1.6   
1  2016-03-24 10:58:45                               A5_Sportback_2.7_Tdi   
2  2016-03-14 12:52:21                     Jeep_Grand_Cherokee_"Overland"   
3  2016-03-17 16:54:04                                 GOLF_4_1_4__3TÜRER   
4  2016-03-31 17:25:20                     Skoda_Fabia_1.4_TDI_PD_Classic   
5  2016-04-04 17:36:23  BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...   
6  2016-04-01 20:48:51                        Peugeot_206_CC_110_Platinum   
7  2016-03-21 18:54:38                       VW_Derby_Bj_80__Scheunenfund   
8  2016-04-04 23:42:13               Ford_C___Max_Titanium_1_0_L_EcoBoost   
9  2016-03-17 10:53:50  VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...   

   seller offerType  price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot    480     test         NaN                1993   
1  privat   Angebot  18300     test       coupe                2011   
2  privat   Angebot   9800     test         suv                2004   
3  privat   Angebot   1500     test  kleinwagen                2001   
4  privat   Angebot   3600     test  kleinwagen                2008   
5  privat   Angebot    650     test   limousine                1995   
6  privat   Angebot   2200     test      cabrio                2004   
7  privat   Angebot      0     test   limousine                1980   
8  privat   Angebot  14500  control         bus                2014   
9  privat   Angebot    999     test  kleinwagen                1998   

     gearbox  powerPS    model  kilometer  monthOfRegistration fuelType  \
0    manuell        0     golf     150000                    0   benzin   
1    manuell      190      NaN     125000                    5   diesel   
2  automatik      163    grand     125000                    8   diesel   
3    manuell       75     golf     150000                    6   benzin   
4    manuell       69    fabia      90000                    7   diesel   
5    manuell      102      3er     150000                   10   benzin   
6    manuell      109  2_reihe     150000                    8   benzin   
7    manuell       50   andere      40000                    7   benzin   
8    manuell      125    c_max      30000                    8   benzin   
9    manuell      101     golf     150000                    0      NaN   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0  volkswagen               NaN  2016-03-24 00:00:00             0   
1        audi                ja  2016-03-24 00:00:00             0   
2        jeep               NaN  2016-03-14 00:00:00             0   
3  volkswagen              nein  2016-03-17 00:00:00             0   
4       skoda              nein  2016-03-31 00:00:00             0   
5         bmw                ja  2016-04-04 00:00:00             0   
6     peugeot              nein  2016-04-01 00:00:00             0   
7  volkswagen              nein  2016-03-21 00:00:00             0   
8        ford               NaN  2016-04-04 00:00:00             0   
9  volkswagen               NaN  2016-03-17 00:00:00             0   

   postalCode             lastSeen  
0       70435  2016-04-07 03:16:57  
1       66954  2016-04-07 01:46:50  
2       90480  2016-04-05 12:47:46  
3       91074  2016-03-17 17:40:17  
4       60437  2016-04-06 10:17:21  
5       33775  2016-04-06 19:17:07  
6       67112  2016-04-05 18:18:39  
7       19348  2016-03-25 16:47:58  
8       94505  2016-04-04 23:42:13  
9       27472  2016-03-31 17:17:06

## 2. Data Cleaning

In [7]:
autos['notRepairedDamage'].value_counts()

notRepairedDamage
nein    263182
ja       36286
Name: count, dtype: int64

We can observe that most of the columns have non null values. However we have to deal with some columns such as 'vehicleType', 'gearbox', 'model', 'fuelType' and 'notRepairedDamage'. 
- For all the columns except 'notRepairedDamage' we will eliminate the rows with null values as these attributes play an important role in deciding which car to buy.
- For 'notRepairedDamage' we can observe that the column can take one out of two values: 'ja' and 'nein'. 'nein' being the highly dominant value in this column we will map for all the NaN values as 'Nein' as well. 

In [3]:
autos['notRepairedDamage'] = autos['notRepairedDamage'].fillna('nein') #converting NaN values to 'nein'
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    371528 non-nu

In [13]:
autos.dropna(axis = 0) #removing all the rows with NaN values.
autos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 299864 entries, 2 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          299864 non-null  object
 1   name                 299864 non-null  object
 2   seller               299864 non-null  object
 3   offerType            299864 non-null  object
 4   price                299864 non-null  int64 
 5   abtest               299864 non-null  object
 6   vehicleType          299864 non-null  object
 7   yearOfRegistration   299864 non-null  int64 
 8   gearbox              299864 non-null  object
 9   powerPS              299864 non-null  int64 
 10  model                299864 non-null  object
 11  kilometer            299864 non-null  int64 
 12  monthOfRegistration  299864 non-null  int64 
 13  fuelType             299864 non-null  object
 14  brand                299864 non-null  object
 15  notRepairedDamage    299864 non-null  o

In [16]:
#cleaning the column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [18]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer(kms)', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen']
autos.head(10)

date_crawled                                               name  \
2   2016-03-14 12:52:21                     Jeep_Grand_Cherokee_"Overland"   
3   2016-03-17 16:54:04                                 GOLF_4_1_4__3TÜRER   
4   2016-03-31 17:25:20                     Skoda_Fabia_1.4_TDI_PD_Classic   
5   2016-04-04 17:36:23  BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...   
6   2016-04-01 20:48:51                        Peugeot_206_CC_110_Platinum   
7   2016-03-21 18:54:38                       VW_Derby_Bj_80__Scheunenfund   
8   2016-04-04 23:42:13               Ford_C___Max_Titanium_1_0_L_EcoBoost   
10  2016-03-26 19:54:18                                  Mazda_3_1.6_Sport   
11  2016-04-07 10:06:22      Volkswagen_Passat_Variant_2.0_TDI_Comfortline   
12  2016-03-15 22:49:09                  VW_Passat_Facelift_35i__"7Sitzer"   

    seller offer_type  price   abtest vehicle_type  registration_year  \
2   privat    Angebot   9800     test          suv               2004   
3   privat    Angebot   1500     test   kleinwagen               2001   
4   privat    Angebot   3600     test   kleinwagen               2008   
5   privat    Angebot    650     test    limousine               1995   
6   privat    Angebot   2200     test       cabrio               2004   
7   privat    Angebot      0     test    limousine               1980   
8   privat    Angebot  14500  control          bus               2014   
10  privat    Angebot   2000  control    limousine               2004   
11  privat    Angebot   2799  control        kombi               2005   
12  privat    Angebot    999  control        kombi               1995   

      gearbox  power_ps    model  odometer(kms)  registration_month fuel_type  \
2   automatik       163    grand         125000                   8    diesel   
3     manuell        75     golf         150000                   6    benzin   
4     manuell        69    fabia          90000                   7    diesel   
5     manuell       102      3er         150000                  10    benzin   
6     manuell       109  2_reihe         150000                   8    benzin   
7     manuell        50   andere          40000                   7    benzin   
8     manuell       125    c_max          30000                   8    benzin   
10    manuell       105  3_reihe         150000                  12    benzin   
11    manuell       140   passat         150000                  12    diesel   
12    manuell       115   passat         150000                  11    benzin   

         brand unrepaired_damage           ad_created  no_of_pictures  \
2         jeep              nein  2016-03-14 00:00:00               0   
3   volkswagen              nein  2016-03-17 00:00:00               0   
4        skoda              nein  2016-03-31 00:00:00               0   
5          bmw                ja  2016-04-04 00:00:00               0   
6      peugeot              nein  2016-04-01 00:00:00               0   
7   volkswagen              nein  2016-03-21 00:00:00               0   
8         ford              nein  2016-04-04 00:00:00               0   
10       mazda              nein  2016-03-26 00:00:00               0   
11  volkswagen                ja  2016-04-07 00:00:00               0   
12  volkswagen              nein  2016-03-15 00:00:00               0   

    postal_code            last_seen  
2         90480  2016-04-05 12:47:46  
3         91074  2016-03-17 17:40:17  
4         60437  2016-04-06 10:17:21  
5         33775  2016-04-06 19:17:07  
6         67112  2016-04-05 18:18:39  
7         19348  2016-03-25 16:47:58  
8         94505  2016-04-04 23:42:13  
10        96224  2016-04-06 10:45:34  
11        57290  2016-04-07 10:25:17  
12        37269  2016-04-01 13:16:16

In [15]:
autos.describe()


price  yearOfRegistration        powerPS      kilometer  \
count  3.715280e+05       371528.000000  371528.000000  371528.000000   
mean   1.729514e+04         2004.577997     115.549477  125618.688228   
std    3.587954e+06           92.866598     192.139578   40112.337051   
min    0.000000e+00         1000.000000       0.000000    5000.000000   
25%    1.150000e+03         1999.000000      70.000000  125000.000000   
50%    2.950000e+03         2003.000000     105.000000  150000.000000   
75%    7.200000e+03         2008.000000     150.000000  150000.000000   
max    2.147484e+09         9999.000000   20000.000000  150000.000000   

       monthOfRegistration  nrOfPictures    postalCode  
count        371528.000000      371528.0  371528.00000  
mean              5.734445           0.0   50820.66764  
std               3.712412           0.0   25799.08247  
min               0.000000           0.0    1067.00000  
25%               3.000000           0.0   30459.00000  
50%               6.000000           0.0   49610.00000  
75%               9.000000           0.0   71546.00000  
max              12.000000           0.0   99998.00000

In [14]:
autos.describe(include=object)

dateCrawled         name  seller offerType  abtest  \
count                371528       371528  371528    371528  371528   
unique               280500       233531       2         2       2   
top     2016-03-24 14:49:47  Ford_Fiesta  privat   Angebot    test   
freq                      7          657  371525    371516  192585   

       vehicleType  gearbox   model fuelType       brand notRepairedDamage  \
count       333659   351319  351044   338142      371528            299468   
unique           8        2     251        7          40                 2   
top      limousine  manuell    golf   benzin  volkswagen              nein   
freq         95894   274214   30070   223857       79640            263182   

                dateCreated             lastSeen  
count                371528               371528  
unique                  114               182806  
top     2016-04-03 00:00:00  2016-04-07 06:45:59  
freq                  14450                   17

In [17]:
autos.describe(include="all")

dateCrawled         name  seller offerType         price  \
count                371528       371528  371528    371528  3.715280e+05   
unique               280500       233531       2         2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat   Angebot           NaN   
freq                      7          657  371525    371516           NaN   
mean                    NaN          NaN     NaN       NaN  1.729514e+04   
std                     NaN          NaN     NaN       NaN  3.587954e+06   
min                     NaN          NaN     NaN       NaN  0.000000e+00   
25%                     NaN          NaN     NaN       NaN  1.150000e+03   
50%                     NaN          NaN     NaN       NaN  2.950000e+03   
75%                     NaN          NaN     NaN       NaN  7.200000e+03   
max                     NaN          NaN     NaN       NaN  2.147484e+09   

        abtest vehicleType  yearOfRegistration  gearbox        powerPS  \
count   371528      333659       371528.000000   351319  371528.000000   
unique       2           8                 NaN        2            NaN   
top       test   limousine                 NaN  manuell            NaN   
freq    192585       95894                 NaN   274214            NaN   
mean       NaN         NaN         2004.577997      NaN     115.549477   
std        NaN         NaN           92.866598      NaN     192.139578   
min        NaN         NaN         1000.000000      NaN       0.000000   
25%        NaN         NaN         1999.000000      NaN      70.000000   
50%        NaN         NaN         2003.000000      NaN     105.000000   
75%        NaN         NaN         2008.000000      NaN     150.000000   
max        NaN         NaN         9999.000000      NaN   20000.000000   

         model      kilometer  monthOfRegistration fuelType       brand  \
count   351044  371528.000000        371528.000000   338142      371528   
unique     251            NaN                  NaN        7          40   
top       golf            NaN                  NaN   benzin  volkswagen   
freq     30070            NaN                  NaN   223857       79640   
mean       NaN  125618.688228             5.734445      NaN         NaN   
std        NaN   40112.337051             3.712412      NaN         NaN   
min        NaN    5000.000000             0.000000      NaN         NaN   
25%        NaN  125000.000000             3.000000      NaN         NaN   
50%        NaN  150000.000000             6.000000      NaN         NaN   
75%        NaN  150000.000000             9.000000      NaN         NaN   
max        NaN  150000.000000            12.000000      NaN         NaN   

       notRepairedDamage          dateCreated  nrOfPictures    postalCode  \
count             299468               371528      371528.0  371528.00000   
unique                 2                  114           NaN           NaN   
top                 nein  2016-04-03 00:00:00           NaN           NaN   
freq              263182                14450           NaN           NaN   
mean                 NaN                  NaN           0.0   50820.66764   
std                  NaN                  NaN           0.0   25799.08247   
min                  NaN                  NaN           0.0    1067.00000   
25%                  NaN                  NaN           0.0   30459.00000   
50%                  NaN                  NaN           0.0   49610.00000   
75%                  NaN                  NaN           0.0   71546.00000   
max                  NaN                  NaN           0.0   99998.00000   

                   lastSeen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

- We can see that no_of_pictures is a redundant column. Hence, we will drop this column as it is not useful in our analysis. 
- We will drop all the rows with unrepaired damages as these vehicles are not desirabe buys. 

In [20]:
autos['seller'].value_counts()

seller
privat        371525
gewerblich         3
Name: count, dtype: int64

In [21]:
autos['offerType'].value_counts()

offerType
Angebot    371516
Gesuch         12
Name: count, dtype: int64

In [23]:
autos['notRepairedDamage'].value_counts()

notRepairedDamage
nein    263182
ja       36286
Name: count, dtype: int64

- From the following three columns we will remove the rows with 'gewerblich', 'Gesich' and 'ja' as these values are insignificant or not relevant to us as a buyer. 
- After removing the rows with these vales we will then remove these columns for a simpler dataframe to work with. 

In [25]:
autos = autos.drop(autos[autos['seller'] == 'gewerblich'].index)       #Removes all the rows from the dataframe with a particular value under a particular column.
autos = autos.drop(autos[autos['offerType'] == 'Gesuch'].index)
autos = autos.drop(autos[autos['notRepairedDamage'] == 'ja'].index)
autos.describe(include='all')

dateCrawled         name  seller offerType         price  \
count                335229       335229  335229    335229  3.352290e+05   
unique               260426       211649       1         1           NaN   
top     2016-03-26 22:57:31  Ford_Fiesta  privat   Angebot           NaN   
freq                      6          575  335229    335229           NaN   
mean                    NaN          NaN     NaN       NaN  1.206395e+04   
std                     NaN          NaN     NaN       NaN  6.921279e+05   
min                     NaN          NaN     NaN       NaN  0.000000e+00   
25%                     NaN          NaN     NaN       NaN  1.300000e+03   
50%                     NaN          NaN     NaN       NaN  3.300000e+03   
75%                     NaN          NaN     NaN       NaN  7.800000e+03   
max                     NaN          NaN     NaN       NaN  1.000000e+08   

        abtest vehicleType  yearOfRegistration  gearbox        powerPS  \
count   335229      301266       335229.000000   316555  335229.000000   
unique       2           8                 NaN        2            NaN   
top       test   limousine                 NaN  manuell            NaN   
freq    173804       87115                 NaN   245616            NaN   
mean       NaN         NaN         2004.895397      NaN     116.883014   
std        NaN         NaN           96.641541      NaN     189.880778   
min        NaN         NaN         1000.000000      NaN       0.000000   
25%        NaN         NaN         1999.000000      NaN      71.000000   
50%        NaN         NaN         2004.000000      NaN     107.000000   
75%        NaN         NaN         2008.000000      NaN     150.000000   
max        NaN         NaN         9999.000000      NaN   20000.000000   

         model      kilometer  monthOfRegistration fuelType       brand  \
count   317113  335229.000000        335229.000000   305799      335229   
unique     251            NaN                  NaN        7          40   
top       golf            NaN                  NaN   benzin  volkswagen   
freq     27526            NaN                  NaN   200708       72627   
mean       NaN  124515.003177             5.777764      NaN         NaN   
std        NaN   40687.578235             3.698878      NaN         NaN   
min        NaN    5000.000000             0.000000      NaN         NaN   
25%        NaN  100000.000000             3.000000      NaN         NaN   
50%        NaN  150000.000000             6.000000      NaN         NaN   
75%        NaN  150000.000000             9.000000      NaN         NaN   
max        NaN  150000.000000            12.000000      NaN         NaN   

       notRepairedDamage          dateCreated  nrOfPictures     postalCode  \
count             263176               335229      335229.0  335229.000000   
unique                 1                  113           NaN            NaN   
top                 nein  2016-04-03 00:00:00           NaN            NaN   
freq              263176                13147           NaN            NaN   
mean                 NaN                  NaN           0.0   50862.158047   
std                  NaN                  NaN           0.0   25816.290519   
min                  NaN                  NaN           0.0    1067.000000   
25%                  NaN                  NaN           0.0   30539.000000   
50%                  NaN                  NaN           0.0   49624.000000   
75%                  NaN                  NaN           0.0   71638.000000   
max                  NaN                  NaN           0.0   99998.000000   

                   lastSeen  
count                335229  
unique               167576  
top     2016-04-06 17:46:58  
freq                     16  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [37]:
autos = autos.drop(columns = ["seller", "offerType","notRepairedDamage","nrOfPictures"]) #Removing the columns with a single value
autos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 335229 entries, 0 to 371527
Data columns (total 16 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          335229 non-null  object
 1   name                 335229 non-null  object
 2   price                335229 non-null  int64 
 3   abtest               335229 non-null  object
 4   vehicleType          301266 non-null  object
 5   yearOfRegistration   335229 non-null  int64 
 6   gearbox              316555 non-null  object
 7   powerPS              335229 non-null  int64 
 8   model                317113 non-null  object
 9   kilometer            335229 non-null  int64 
 10  monthOfRegistration  335229 non-null  int64 
 11  fuelType             305799 non-null  object
 12  brand                335229 non-null  object
 13  dateCreated          335229 non-null  object
 14  postalCode           335229 non-null  int64 
 15  lastSeen             335229 non-null  o

With this we finish our data cleaning work and now we can concentrate on the Analysis part of our project. 

## 3. Exploring the odometer and price columns.

Here I will look at odometer and price columns and remove the rows with unrealistically high or low values(outliers) 

In [42]:
autos['price'].describe()

count    3.352290e+05
mean     1.206395e+04
std      6.921279e+05
min      0.000000e+00
25%      1.300000e+03
50%      3.300000e+03
75%      7.800000e+03
max      1.000000e+08
Name: price, dtype: float64

In [44]:
autos['price'].value_counts(ascending = False)

price
0        8767
1500     4542
2500     3965
1200     3907
500      3825
         ... 
3440        1
1997        1
44998       1
1179        1
10985       1
Name: count, Length: 5540, dtype: int64

We have 8767 cars listed with price 0, we will treat this as price not listed and remove the rows

In [46]:
autos = autos.drop(autos[autos['price']==0].index)

In [51]:
autos["price"].value_counts().sort_index(ascending = False).head(50)


price
99999999    14
99000000     1
74185296     1
32545461     1
27322222     1
14000500     1
12345678     7
11111111    10
10010011     1
10000000     6
9999999      2
3895000      1
3890000      1
2995000      1
2795000      1
1600000      2
1300000      1
1250000      2
1234566      1
1111111      2
1010010      1
1000000      3
999999      13
911911       1
849000       1
820000       1
780000       1
745000       2
725000       1
700000       1
650000       1
619000       1
600000       2
599000       1
585000       1
579000       1
517895       1
500000       2
488997       1
487000       1
485000       1
466000       1
445000       1
440000       1
420000       1
399997       1
395000       1
390000       1
370000       2
368000       1
Name: count, dtype: int64

In [52]:
too_cheap = autos['price'].between(0,1000)
too_expensive = autos['price']>20000
extremes = too_cheap | too_expensive 
autos = autos[~extremes]

In [53]:
autos['price'].describe()

count    250357.000000
mean       5827.433629
std        4563.865777
min        1001.000000
25%        2250.000000
50%        4200.000000
75%        8000.000000
max       20000.000000
Name: price, dtype: float64

In [54]:
autos['kilometer'].describe()

count    250357.000000
mean     125236.462332
std       38548.783759
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: kilometer, dtype: float64

In [55]:
autos['kilometer'].value_counts().sort_index(ascending=False).head(50)

kilometer
150000    155979
125000     28531
100000     11954
90000       9800
80000       8733
70000       7680
60000       6778
50000       5721
40000       4567
30000       3861
20000       3148
10000        947
5000        2658
Name: count, dtype: int64

No outliers in the kilometer column 

## 4. Exploring the date columns in the data

In [58]:
date_crawler = autos['dateCrawled'].str[:10]
date_crawler.value_counts(normalize=True, dropna=False).sort_values(ascending=False)

dateCrawled
2016-04-03    0.039967
2016-04-04    0.037886
2016-03-12    0.037287
2016-03-20    0.037015
2016-03-14    0.036484
2016-04-02    0.036176
2016-03-19    0.035793
2016-03-28    0.035529
2016-03-07    0.035042
2016-03-21    0.034898
2016-04-01    0.034339
2016-03-29    0.033632
2016-03-25    0.033249
2016-03-09    0.033205
2016-03-26    0.033017
2016-03-30    0.032893
2016-03-08    0.032785
2016-03-11    0.032745
2016-03-10    0.032613
2016-03-15    0.032090
2016-03-22    0.031902
2016-03-23    0.031651
2016-03-31    0.031251
2016-03-27    0.030824
2016-03-17    0.030812
2016-03-16    0.029769
2016-03-24    0.029678
2016-03-05    0.026047
2016-03-13    0.016177
2016-03-06    0.014871
2016-03-18    0.013089
2016-04-05    0.012530
2016-04-06    0.003163
2016-04-07    0.001590
Name: proportion, dtype: float64

In [59]:
days_last_seen = autos["lastSeen"].str[:10]
days_last_seen.value_counts(normalize=True, dropna=False).sort_values()

lastSeen
2016-03-05    0.001074
2016-03-06    0.003667
2016-03-07    0.004494
2016-03-18    0.006738
2016-03-08    0.007445
2016-03-13    0.007829
2016-03-09    0.008608
2016-03-10    0.010421
2016-03-14    0.011747
2016-03-11    0.012251
2016-03-15    0.014915
2016-03-26    0.015310
2016-03-16    0.015394
2016-03-19    0.015909
2016-03-27    0.016045
2016-03-23    0.017179
2016-03-25    0.018306
2016-03-24    0.018681
2016-03-20    0.018989
2016-03-21    0.019097
2016-03-22    0.019656
2016-03-28    0.021573
2016-03-12    0.022328
2016-03-29    0.022480
2016-03-30    0.022664
2016-03-31    0.023139
2016-04-01    0.023470
2016-04-04    0.024281
2016-04-02    0.024657
2016-04-03    0.024781
2016-03-17    0.027245
2016-04-05    0.132275
2016-04-07    0.136956
2016-04-06    0.230395
Name: proportion, dtype: float64

In [61]:
date_created = autos["dateCreated"].str[:10]
date_created.value_counts(normalize=True, dropna=False).sort_values()

dateCreated
2015-06-18    0.000004
2015-11-10    0.000004
2015-12-17    0.000004
2016-01-03    0.000004
2015-08-10    0.000004
                ...   
2016-04-02    0.036116
2016-03-20    0.037035
2016-03-12    0.037051
2016-04-04    0.037970
2016-04-03    0.040159
Name: proportion, Length: 107, dtype: float64

## 5. Dealing with registration dates

In [65]:
old = autos['yearOfRegistration']<1990
future = autos['yearOfRegistration']>2016
incorrect = old | future
autos = autos[~incorrect]

In [66]:
autos['yearOfRegistration'].describe()

count    233860.000000
mean       2004.423907
std           5.097009
min        1990.000000
25%        2001.000000
50%        2005.000000
75%        2008.000000
max        2016.000000
Name: yearOfRegistration, dtype: float64

In [67]:
autos['yearOfRegistration'].value_counts(normalize=True).sort_values(ascending=False)


yearOfRegistration
2005    0.080860
2006    0.077504
2004    0.073959
2003    0.070739
2007    0.067741
2002    0.063949
2008    0.061562
2001    0.059741
2009    0.058266
2000    0.056175
1999    0.050607
2010    0.043530
2011    0.040272
1998    0.035243
2012    0.029240
1997    0.022069
2016    0.017566
2013    0.016749
1996    0.016206
1995    0.012123
2014    0.009972
1994    0.008001
1992    0.006393
1993    0.005982
1991    0.005824
1990    0.005580
2015    0.004148
Name: proportion, dtype: float64

Here we have dealt with cars that are too old to be bought and also incorrect listing that have year of registration in the future.

## 6. Exploring price by brand 

In this section we explore mean prices of the top car brands. 

In [70]:
brand_count = autos['brand'].value_counts()
print(brand_count)

brand
volkswagen        50799
bmw               28644
mercedes_benz     24426
audi              22353
opel              21568
ford              13926
renault            9264
peugeot            7221
fiat               5029
skoda              4639
seat               4470
smart              4357
toyota             3688
mazda              3569
citroen            3328
nissan             2935
mini               2888
hyundai            2702
volvo              2033
kia                1817
honda              1753
mitsubishi         1590
suzuki             1520
alfa_romeo         1444
sonstige_autos     1320
chevrolet          1253
chrysler            882
dacia               791
jeep                509
porsche             486
land_rover          431
subaru              396
jaguar              384
saab                347
daihatsu            341
daewoo              216
lancia              186
rover               170
lada                136
trabant              49
Name: count, dtype: int64


We shall consider top brand to be those that have atleast 1000 ads posted. 

In [71]:
top_brands = brand_count[brand_count > 1000].index
top_brands

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford', 'renault',
       'peugeot', 'fiat', 'skoda', 'seat', 'smart', 'toyota', 'mazda',
       'citroen', 'nissan', 'mini', 'hyundai', 'volvo', 'kia', 'honda',
       'mitsubishi', 'suzuki', 'alfa_romeo', 'sonstige_autos', 'chevrolet'],
      dtype='object', name='brand')

In [3]:
brand_price = {}

for brand in top_brands:
    mean = autos[autos['brand'] == brand]['price'].mean()
    brand_price[brand] = round(mean, 2)
    
brand_price
    

NameError: name 'top_brands' is not defined

In [84]:
brand_price_series = pd.Series(brand_price)
brand_price_df = pd.DataFrame(brand_price_series, columns= ["mean_price"])
brand_price_df.sort_values(by='mean_price', ascending = False)

mean_price
mini               9477.26
audi               7566.46
sonstige_autos     7305.80
bmw                7272.65
mercedes_benz      6832.86
skoda              6552.53
hyundai            6235.00
kia                6151.33
nissan             6100.41
volvo              5956.89
chevrolet          5892.45
volkswagen         5814.20
seat               5484.80
toyota             5380.29
suzuki             5223.01
honda              5130.55
mazda              5043.47
ford               4814.27
mitsubishi         4691.34
alfa_romeo         4638.68
citroen            4461.87
opel               4220.04
peugeot            4182.29
fiat               4042.90
smart              3942.57
renault            3590.28

In [90]:
brand_mileage = {}

for brand in brand_price_df.index[:10]:
    mean = autos[autos['brand'] == brand]['kilometer'].mean()
    brand_mileage[brand] = round(mean, 2)
    
print(brand_mileage)

{'volkswagen': 128441.41, 'bmw': 137259.64, 'mercedes_benz': 135104.4, 'audi': 135463.25, 'opel': 123355.43, 'ford': 120712.7, 'renault': 120539.72, 'peugeot': 119035.45, 'fiat': 107489.56, 'skoda': 113923.26}


In [93]:
brand_mileage_series = pd.Series(brand_mileage)
brand_price_df['avg_mileage'] = brand_mileage_series
final_df = brand_price_df[:10]
print(final_df)

               mean_price  avg_mileage
volkswagen        5814.20    128441.41
bmw               7272.65    137259.64
mercedes_benz     6832.86    135104.40
audi              7566.46    135463.25
opel              4220.04    123355.43
ford              4814.27    120712.70
renault           3590.28    120539.72
peugeot           4182.29    119035.45
fiat              4042.90    107489.56
skoda             6552.53    113923.26


In [95]:
final_df = final_df.sort_values(by='mean_price', ascending = False)
print(final_df)

               mean_price  avg_mileage
audi              7566.46    135463.25
bmw               7272.65    137259.64
mercedes_benz     6832.86    135104.40
skoda             6552.53    113923.26
volkswagen        5814.20    128441.41
ford              4814.27    120712.70
opel              4220.04    123355.43
peugeot           4182.29    119035.45
fiat              4042.90    107489.56
renault           3590.28    120539.72


## 7. Conclusion

- Here, we can notice that the brands with higher mileage tend to have a higher mean price. However, this is not statistically significant as the brand value also makes a difference. The higly priced brand : audi, bmw, mercedes_benz are widely reknown as the luxury vehciles.
- This final table gives the buyer a good outlook of the data. Based on this table the buyer can shortlist the brand(s) they would like to buy, filter the dataset to show those specific brand(s) which would make the decision of which car to buy a lot more simple.